tuplex/core/src/physical/ExceptionSourceTaskBuilder.cc

```diff
@@ -0,0 +1,294 @@
+            auto pip_res = PipelineBuilder::call(builder, processRowFunc, tuple, userData, builder.CreateLoad(rowNumberVar), initIntermediate(builder));
+
+            // create if based on resCode to go into exception block
+            auto ecCode = builder.CreateZExtOrTrunc(pip_res.resultCode, env().i64Type());
+            auto ecOpID = builder.CreateZExtOrTrunc(pip_res.exceptionOperatorID, env().i64Type());
+            auto numRowsCreated = builder.CreateZExtOrTrunc(pip_res.numProducedRows, env().i64Type());
+
+            // add number of rows created to output row number variable
+            auto outputRowNumber = builder.CreateLoad(rowNumberVar);
+            builder.CreateStore(builder.CreateAdd(builder.CreateLoad(rowNumberVar), numRowsCreated), rowNumberVar);
+
+            auto exceptionRaised = builder.CreateICmpNE(ecCode, env().i64Const(ecToI32(ExceptionCode::SUCCESS)));
+
+            llvm::BasicBlock* bbPipelineFailedUpdate = llvm::BasicBlock::Create(context, "pipeline_failed", builder.GetInsertBlock()->getParent());
+            llvm::BasicBlock* bbPipelineOK = llvm::BasicBlock::Create(context, "pipeline_ok", builder.GetInsertBlock()->getParent());
+            llvm::BasicBlock* curBlock = builder.GetInsertBlock();
+            llvm::BasicBlock* bbPipelineFailed = exceptionBlock(builder, userData, ecCode, ecOpID, outputRowNumber, inputRowPtr, inputRowSize); // generate exception block (incl. ignore & handler if necessary)
+
+
+            llvm::BasicBlock* lastExceptionBlock = builder.GetInsertBlock();
+            llvm::BasicBlock* bbPipelineDone = llvm::BasicBlock::Create(context, "pipeline_done", builder.GetInsertBlock()->getParent());
+
+            builder.SetInsertPoint(curBlock);
+            builder.CreateCondBr(exceptionRaised, bbPipelineFailedUpdate, bbPipelineOK);
+
+            builder.SetInsertPoint(bbPipelineFailedUpdate);
+            builder.CreateStore(builder.CreateAdd(builder.CreateLoad(badRowCountVar), env().i64Const(1)), badRowCountVar);
+            builder.CreateBr(bbPipelineFailed);
+
+            // pipeline ok
+            builder.SetInsertPoint(bbPipelineOK);
+            llvm::Value *normalRowCount = builder.CreateLoad(normalRowCountVar, "normalRowCount");
+            builder.CreateStore(builder.CreateAdd(normalRowCount, env().i64Const(1)), normalRowCountVar);
+            builder.CreateBr(bbPipelineDone);
+
+            // connect exception block to pipeline failure
+            builder.SetInsertPoint(lastExceptionBlock);
+            builder.CreateBr(bbPipelineDone);
+
+            builder.SetInsertPoint(bbPipelineDone);
+
+            // call runtime free all
+            _env->freeAll(builder);
+        }
+
+        void ExceptionSourceTaskBuilder::createMainLoop(llvm::Function *read_block_func) {
+            using namespace std;
+            using namespace llvm;
+
+            assert(read_block_func);
+
+            auto& context = env().getContext();
```

Comment:
can we add some comment on what this code does? Hard to understand without comment. You can also insert some pseudocode/C++ code describing what is genreated here.